### PR TITLE
`.ci/merge-fixes.sh`: Obtain patches via URL, make customizable by repository variable

### PR DESCRIPTION
**Pull request by mkoeppe**
Instead of using `gh pr checkout`, we obtain the CI fixes via their patch URLs.

- This is faster because typically we do not have to unshallow the repo to apply the patches (seconds instead of ~2 minutes)
- Fewer surprises when applied to a PR based on an older release
- Conjecturally uses fewer API queries, helping avoid #36685

When a repository variable `SAGE_CI_FIXES_FROM_REPOSITORIES` is set in a fork, it is used instead of the hardcoded sagemath/sage as the source(s) of the CI fixes; this gives better control in decentralized development. When set to "none", this also makes it possible to see the "ground truth", addressing a concern raised in https://github.com/sagemath/sage/pull/36349#issuecomment-1738400399. See comments at the top of `.ci/merge-fixes.sh` for details.

Also improving the display of details of the merged PRs, as requested by @tornaria in https://github.com/sagemath/sage/pull/36696#issuecomment-1806916614

Example runs:
- https://github.com/sagemath/sage/actions/runs/6854931832/job/18638901859?pr=36686 (here it recovers gracefully from failed merges due to the 10.2.rc1/10.2.rc2 tagging confusions)
- https://github.com/mkoeppe/sage/actions/runs/6855410957/job/18640380398#step:3:8 (I have set `SAGE_CI_FIXES_FROM_REPOSITORIES=none` in my fork -- for testing this PR)

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: .ci/merge-fixes.sh [hunk 1]
```diff
@@ -2,7 +2,9 @@
 # Merge open PRs from sagemath/sage labeled "blocker".
 REPO="sagemath/sage"
 GH="gh -R $REPO"
-PRs="$($GH pr list --label "p: blocker / 1" --json number --jq '.[].number')"
+mkdir -p upstream
+PRs="$($GH pr list --label "p: blocker / 1" --json number --jq '.[].number' | tee upstream/ci-fixes.txt)"
```
Comment (tornaria) [inline .ci/merge-fixes.sh]: Why not add `--label "s: positive review"`? Multiple `--label` arguments work just fine.

File: .ci/merge-fixes.sh [hunk 2]
```diff
@@ -14,23 +16,42 @@ else
...
+        # Here, we obtain the "git am"-able patch of the PR branch.
+        # This also makes it unnecessary to unshallow the repository.
+        #
+        # Considered alternative: Use https://github.com/$REPO/pull/$a.diff,
+        # which squashes everything into one diff without commit metadata.
+        PATH=build/bin:$PATH build/bin/sage-download-file "https://github.com/$REPO/pull/$a.patch" upstream/$a.patch
+        date -u +"%Y-%m-%dT%H:%M:%SZ" > upstream/$a.date  # Record the date, for future reference
+        LAST_SHA=$(sed -n -E '/^From [0-9a-f]{40}/s/^From ([0-9a-f]{40}).*/\1/p' upstream/$a.patch | tail -n 1)
+        echo "::group::Applying PR https://github.com/$REPO/pull/$a @ https://github.com/$REPO/commit/$LAST_SHA as a patch"
+        export GIT_COMMITTER_NAME="$GIT_AUTHOR_NAME applying https://github.com/$REPO/pull/$a @ https://github.com/$REPO/commit/$LAST_SHA"
```
Comment (tornaria) [inline .ci/merge-fixes.sh]: I realized `https://github.com/$REPO/commits/$LAST_SHA` is even better.

Compare:
 - https://github.com/sagemath/sage/commit/4a70c1e535bc621bdbe95d8f531375e5700d7c3e
 - https://github.com/sagemath/sage/commits/4a70c1e535bc621bdbe95d8f531375e5700d7c3e

**Review by tornaria — COMMENTED** [review]
LGTM in general.

File: .ci/merge-fixes.sh (hunk 2 above)
Comment (mkoeppe) [reply .ci/merge-fixes.sh]: Does that work in logs?

File: .ci/merge-fixes.sh (hunk 1 above)
Comment (mkoeppe) [reply .ci/merge-fixes.sh]: We can do that later when the label-warring phase has been outgrown

File: .ci/merge-fixes.sh (hunk 2 above)
Comment (mkoeppe) [reply .ci/merge-fixes.sh]: I want to be careful there so that when the workflow runs in a fork, these links still correctly point to sagemath/sage

Comment (tornaria) [reply .ci/merge-fixes.sh]: In the logs only full URLs are clickable, and there is no translation like here. The point is not that it looks shorter, but that the url `https://github.com/sagemath/sage/commit/4a70c1e535bc621bdbe95d8f531375e5700d7c3e` that you are using now is a little bit less nice than the url `https://github.com/sagemath/sage/commits/4a70c1e535bc621bdbe95d8f531375e5700d7c3e`. From the latter one can click on the individual commit to see the patch but it's nicer as a summary of the PR changes when there's more than one commit.

Comment (mkoeppe) [reply .ci/merge-fixes.sh]: Ah thanks, I missed that from how github formatted your comment https://github.com/sagemath/sage/pull/36686#discussion_r1390519957

Comment (mkoeppe) [reply .ci/merge-fixes.sh]: Done in 9dac7295709.

Comment (tornaria) [reply .ci/merge-fixes.sh]: The links still work using `sagemath/sage` as the repo even if the commit doesn't belong to that repo but a fork.

**Comment by tornaria** [comment]
LGTM, but how is it tested?

**Comment by mkoeppe** [comment]
So far, only on its own CI run...

**Comment by mkoeppe** [comment]
I've reworked and generalized this a bit. Ready for review.

**Comment by kwankyu** [comment]
> LGTM, but how is it tested?

See it in action here:

https://github.com/kwankyu/sage/actions/runs/6887591924/job/18735031877

It seems working well. You can even watch it recovering from applying-patch failure.

**Comment by mkoeppe** [comment]
Thanks for testing!

> Comparing with the original `merge-fixes.sh` [...]
> 
> (1) It is just slightly faster.

Note it is much faster in runs where patches apply directly

> (2) It fails in applying the blocker #36617 but the original `merge-fixes.sh` succeeds.

I think this failure would go away if we rebased #36617.

**Comment by kwankyu** [comment]
> > Comparing with the original `merge-fixes.sh` [...]
> > (1) It is just slightly faster.
> 
> Note it is much faster in runs where patches apply directly

Right. It was slow this time because of #36617 that failed to be applied directly.

> > (2) It fails in applying the blocker #36617 but the original `merge-fixes.sh` succeeds.
> 
> I think this failure would go away if we rebased #36617.

Failing in applying non-rebased blockers may be a good thing, by the way.

**Review by kwankyu — APPROVED** [review]
LGTM too.

**Comment by mkoeppe** [comment]
> Failing in applying non-rebased blockers may be a good thing, by the way.

I think so, too. It sort of enforces some restraint on what to do in these CI fixes.

**Comment by mkoeppe** [comment]
Thanks both for the review!

**Comment by tobiasdiez** [comment]
It fails to apply https://github.com/sagemath/sage/pull/36372 now, although this PR is up-to-date with the newest dev branch. It's seems like an unnecessary constraint that blocker PR's need to be rebased (instead of merging the latest develop).

**Pull request merged**